src/perform_step/rkc_perform_step.jl

```diff
@@ -153,77 +162,87 @@ end
   @unpack ms, fpa, fpb, fpbe, recf = cache
   maxeig!(integrator, cache)
   # The the number of degree for Chebyshev polynomial
-  mdeg = Int(floor(sqrt((3 + dt * integrator.eigen_est)/0.353) + 1))
+  mdeg = Int(floor(sqrt((3 + dt*integrator.eigen_est)/0.353) + 1))
   mdeg = min(max(mdeg,cache.min_stage), cache.max_stage)
   cache.mdeg = max(mdeg, 5) - 4
   choosedeg!(cache)
   # recurrence
   # for the first stage
-  temp1 = dt * recf[cache.recind]
-  ci1 = t + temp1
-  ci2 = t + temp1
-  ci3 = t
-  gprev2 = copy(uprev)
-  gprev = uprev + temp1 * fsalfirst
-  ms[cache.mdeg] < 2 && ( u = gprev )
-  # for the second to the ms[cache.mdeg] th stages
-  for i in 2:ms[cache.mdeg]
-    μ, κ = recf[cache.recind + (i - 2)*2 + 1], recf[cache.recind + (i - 2)*2 + 2]
-    ν = -1 - κ
-    dtμ = dt*μ
-    u = f(gprev, p,ci1)
-    ci1 = dtμ - ν * ci2 - κ * ci3
-    u = dtμ * u - ν * gprev - κ * gprev2
-    i < ms[cache.mdeg] && (gprev2 = gprev; gprev = u)
-    ci3 = ci2
-    ci2 = ci1
+  tᵢ₋₁ = t + dt*recf[cache.start]
+  tᵢ₋₂ = t + dt*recf[cache.start]
+  tᵢ₋₃ = t
+  uᵢ₋₂ = copy(uprev)
+  uᵢ₋₁ = uprev + (dt*recf[cache.start])*fsalfirst
+  cache.mdeg < 2 && ( u = uᵢ₋₁ )
+  # for the second to the cache.mdeg th stages
+  for i in 2:cache.mdeg
+    μ, κ = recf[cache.start + (i - 2)*2 + 1], recf[cache.start + (i - 2)*2 + 2]
+    ν = - 1 - κ
+    u = f(uᵢ₋₁, p,tᵢ₋₁)
+    tᵢ₋₁ = dt*μ - ν*tᵢ₋₂ - κ*tᵢ₋₃
+    u = (dt*μ)*u - ν*uᵢ₋₁ - κ*uᵢ₋₂
+    i < cache.mdeg && (uᵢ₋₂ = uᵢ₋₁; uᵢ₋₁ = u)
+    tᵢ₋₃ = tᵢ₋₂
+    tᵢ₋₂ = tᵢ₋₁
   end
+
+  # These constants correspond to the Buther Tableau coefficients of explicit RK methods
+  a₂₁ = dt*fpa[cache.deg_index][1]
+  a₃₁ = dt*fpa[cache.deg_index][2]; a₃₂ = dt*fpa[cache.deg_index][3]
+  a₄₁ = dt*fpa[cache.deg_index][4]; a₄₂ = dt*fpa[cache.deg_index][5]; a₄₃ = dt*fpa[cache.deg_index][6]
+  B₁  = dt*fpb[cache.deg_index][1]; B₂  = dt*fpb[cache.deg_index][2]; B₃  = dt*fpb[cache.deg_index][3]; B₄  = dt*fpb[cache.deg_index][4]
+  # coefficients of embedded method for error estimation
+  B̂₁  = dt*(fpbe[cache.deg_index][1] - fpb[cache.deg_index][1])
+  B̂₂  = dt*(fpbe[cache.deg_index][2] - fpb[cache.deg_index][2])
+  B̂₃  = dt*(fpbe[cache.deg_index][3] - fpb[cache.deg_index][3])
+  B̂₄  = dt*(fpbe[cache.deg_index][4] - fpb[cache.deg_index][4])
+  B̂₅  = dt*fpbe[cache.deg_index][5]
+
   # 4-stage finishing procedure.
   # Stage-1
-  temp1 = dt * fpa[cache.mdeg][1]
-  gprev = f(u, p, ci1)
+  uᵢ₋₁ = f(u, p, tᵢ₋₁)
   integrator.destats.nf += 1
-  gprev3 = u + temp1 * gprev
+  uᵢ₋₂ = u + a₃₁*uᵢ₋₁
+  uᵢ₋₃ = u + a₄₁*uᵢ₋₁
+  u    += B₁*uᵢ₋₁
+  integrator.opts.adaptive && (tmp = B̂₁*uᵢ₋₁)
+  uᵢ₋₁ = u + (a₂₁ - B₁)*uᵢ₋₁
```

Comment:
This is addition subtraction of the same thing but the only way to avoid it is to add another array. Also both `a21` and `B1` are constants so I think doing this should not be an issue.

Reply:
yeah this should be fine.